# STA130 Week 01 Homework 

Please see the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) for the list of topics covered in this homework assignment, and a list of topics that might appear during ChatBot conversations which are "out of scope" for the purposes of this homework assignment (and hence can be safely ignored if encountered)


<details class="details-example"><summary style="color:blue"><u>Introduction</u></summary>

### Introduction
    
A reasonable characterization of STA130 Homework is that it simply defines a weekly reading comprehension assignment. 
Indeed, STA130 Homework essentially boils down to completing various understanding confirmation exercises oriented around coding and writing tasks.
However, rather than reading a textbook, STA130 Homework is based on ChatBots so students can interactively follow up to clarify questions or confusion that they may still have regarding learning objective assignments.

> Communication is a fundamental skill underlying statistics and data science, so STA130 Homework based on ChatBots helps practice effective two-way communication as part of a "realistic" dialogue activity supporting underlying conceptual understanding building. 

It will likely become increasingly tempting to rely on ChatBots to "do the work for you". But when you find yourself frustrated with a ChatBots inability to give you the results you're looking for, this is a "hint" that you've become overreliant on the ChatBots. Your objective should not be to have ChatBots "do the work for you", but to use ChatBots to help you build your understanding so you can efficiently leverage ChatBots (and other resources) to help you work more efficiently.<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Instructions</u></summary>

### Instructions
    
1. Code and write all your answers (for both the "Prelecture" and "Postlecture" HW) in a python notebook (in code and markdown cells) 
    
> It is *suggested but not mandatory* that you complete the "Prelecture" HW prior to the Monday LEC since (a) all HW is due at the same time; but, (b) completing some of the HW early will mean better readiness for LEC and less of a "procrastentation cruch" towards the end of the week...
    
2. Paste summaries of your ChatBot sessions (including link(s) to chat log histories if you're using ChatGPT) within your notebook
    
> Create summaries of your ChatBot sessions by using concluding prompts such as "Please provide a summary of our exchanges here so I can submit them as a record of our interactions as part of a homework assignment" or, "Please provide me with the final working verson of the code that we created together"
    
3. Save your python jupyter notebook in your own account and "repo" on [github.com](github.com) and submit a link to that notebook though Quercus for assignment marking<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Prompt Engineering?</u></summary>

### Prompt Engineering?    
    
The questions (as copy-pasted prompts) are designed to initialize appropriate ChatBot conversations which can be explored in the manner of an interactive and dynamic textbook; but, it is nonetheless **strongly recommendated** that your rephrase the questions in a way that you find natural to ensure a clear understanding of the question. Given sensible prompts the represent a question well, the two primary challenges observed to arise from ChatBots are 

1. conversations going beyond the intended scope of the material addressed by the question; and, 
2. unrecoverable confusion as a result of sequential layers logial inquiry that cannot be resolved. 

In the case of the former (1), adding constraints specifying the limits of considerations of interest tends to be helpful; whereas, the latter (2) is often the result of initial prompting that leads to poor developments in navigating the material, which are likely just best resolve by a "hard reset" with a new initial approach to prompting.  Indeed, this is exactly the behavior [hardcoded into copilot](https://answers.microsoft.com/en-us/bing/forum/all/is-this-even-normal/0b6dcab3-7d6c-4373-8efe-d74158af3c00)...

</details>


### Marking Rubric (which may award partial credit) 

- [0.1 points]: All relevant ChatBot summaries [including link(s) to chat log histories if you're using ChatGPT] are reported within the notebook
- [0.2 points]: Reasonable well-written general definitions for Question "2.2"
- [0.3 points]: Demonstrated understanding regarding Question "4"
<!-- - [0.2 points]: A sensible justification for the choice in Question "7.4" -->
- [0.4 points]: Requested assessment of ChatBot versus google performance in Question "8.3"


## "Prelecture" HW [*completion prior to next LEC is suggested but not mandatory*]

#### 1. Pick one of the datasets from the ChatBot session(s) of the **TUT demo** (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If your TA has not shared a relevant ChatBot session from their **TUT demo** through a piazza post and a Quercus announcement, the **TUT notebook** has links to example ChatBot sessions that you can use; or, ...
> 
> ```python
> # feel free to just use the following if you prefer...
> import pandas as pd
> url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
> df = pd.read_csv(url)
> df.isna().sum()
> ```
    
</details>

#### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a `pandas` DataFrame has, and then

1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,  
2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> A good initial prompt to start would be be something like
> - "I've downloaded a dataset about characters from animal crossings (from https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv), and I'd like to know what columns of information I have and how much data I have"
> 
> You can further reduce the scope of your inquiry with if needed with something like
> - "I've already downloaded the data and want to understand the size (or dimensions) of the dataset to start with"
> 
> *Some ChatBots can upload your data and do this for you; but, extended usage of this feature [likely requires a paid subscription](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00006_gpt3p5_LoadDataPaywall.md); and, anyway, you need to run the code yourself rather than having a ChatBot do that for you; and, for STA130 we don't want a ChatBot to just do the analysis for us; rather, we instead want ChatBots to help us understand the steps we need to take to analyze the data; so,* **you DO NOT need to purchase an upgraded version of any ChatBots**
> - Free-tier level ChatBots like [GPT4o-mini](https://chat.openai.com/) or [Copilot](https://copilot.microsoft.com/) (which is partially based on [ChatGPT4.0](https://chat.openai.com/), and which you have access to through your UofT account) are sufficiently sophisticated and perfectly appropriate for the STA130 course
    
</details>

#### 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Use your ChatBot session to help you create working examples of using  `df.describe()` and `df['column'].value_counts()` for your dataset (although note that the `.value_counts()` method is not really meant to be used for numeric variables, so if you dataset has only numeric variables, `.value_counts()` might not be particularly informative...)
>
> #### ChatBot Response Scope 
>     
> If prompts are not sufficiently focused you will likely get overly broad responses from the ChatBot, but you can always respond with subsequent refinement requests to appropriately limit the scope of the ChatBot responses to focus on addressing your actual content targets; so, 
> - an initially very general inquiry like, "I need help analyzing my data" will likely result in a ChatBot response suggesting a wide variety of approaches and techniques for summarizing your dataset; but, re-prompting the ChatBot with something like, "What's the simplest form of summarization of this dataset that I could do and how do I do it in Python?" or suggesting guidance using the specific summarization methods requested above will helpfully re-orient the ChatBot to your specific interests and needs
> 
> #### Jupyter Notebook Hints
> 
> Jupyter notebook printouts usaully don't show all of the data (when there's too much to show, like if `df.describe()` includes results for many columns), but the printouts just show enough of the data to give an idea of what the results are which is all we're looking for at the moment
> 
> - Consider dividing the code that ChatBot provides you into different jupyter notebook cells so that each cell concludes with a key printed result; the last line of code in a jupyter notebook cell will automatically print out in a formatted manner, so replacing something like `print(df.head())` with `df.head()` as the last line of a cell provides a sensible way to organize your code
> - The printout suggestions above are demonstrated in `STA130F24_CourseProject.ipynb` if looking at an example would be helpful to understand what they're getting at...
    
</details>
                
#### 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by `df.shape` and what is reported by `df.describe()` with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If the dataset you're using does not have (a) non-numeric variables and (b) missing values in numeric variables (e.g., the `"villagers.csv"` example above has only a single numeric variable `row_n` which has no missing values), instead download and use the [https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv)" data to answer this question  
>
> In (a) above, the "columns it analyzes" refers to the columns of the output of `df.describe()` which will only include "numeric" columns by default, but you can can see the names of all the columns in a dataset using `df.columns`; and, make sure `df.shape` is refering to the dataset you think it is... if you've loaded a different dataset it might not have been called `df`(!)
>
> **If you get any errors (for example related to column names), copy and paste them as a response to the ChatBot, and see if it can help you resove them by adding the suggested adjustments to your code and then reruning all your code to see if the changes have fixed the problem (and repeat this process as needed until the problems have been resolved).**
    
</details>
    
#### 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

- an "attribute", such as `df.shape` which does not end with `()`
- and a "method", such as `df.describe()` which does end with `()` 
   

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The fact that a "method" such as `df.describe()` ends with `()` suggests that "methods" are essentially something that we would call a "function" in programming language terminology; but, without getting too technical or "in the weeds", it might also be worth considering that we could also contrast what the difference is between a "function" in a programming language versus a "function" in mathematics...  
    
</details><br><br>

***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***<br><br>

<details class="details-example"><summary style="color:blue"><u>Continue now...?</u></summary>

### Prelecture VS Postlecture HW
    
Feel free to work on the "Postlecture" HW below if you're making good progress and want to continue: in this case this is particularly reasonable as questions "6" and "7" below directly follow up and extend the "Prelecture" HW questions

*The benefits of continue would are that (a) it might be fun to try to tackle the challenge of working through some problems without additional preparation or guidance; and (b) this is a very valable skill to be comfortable with; and (c) it will let you build experience interacting with ChatBots (and beginning to understand their strengths and limitations in this regard)... it's good to have sense of when using a ChatBot is the best way to figure something out, or if another approach (such as course provided resources or a plain old websearch for the right resourse) would be more effective*
    
</details>    


## "Postlecture" HW [*submission along with "Prelecture" HW is due prior to next TUT*]

#### 6. The `df.describe()` method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The answers here actually make it obvious why these can only be calculated for numeric variables in a dataset, which should help explain the answer to "4(a)" and "4(b)" above
>   
> Also notice that when `df.describe()` is used missing values are not explicitly removed, but `df.describe()`  provides answers anyway. Is it clear what `df.describe()` does with the data in each columns it analyzes if there is missing data in the column in question? 
>
> The next questions addresses removing rows or columns from a dataset in order to explicitly remove the presense of any missingness in the dataset (assuming we're not going to fill in any missing data values using any missing data imputation methods, which are beyond the scope of STA130); so, the behavior of `df.describe()` hints that explicitly removing missing may not always be necessary; but, the concern, though, is that not all methods may be able to handle missing data the way `df.describe()` does...
    
</details>

#### 7. Missing data can be considered "across rows" or "down columns".  Consider how `df.dropna()` or `del df['col']` should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

1. Provide an example of a "use case" in which using `df.dropna()` might be peferred over using `del df['col']`<br><br>
    
2. Provide an example of "the opposite use case" in which using `del df['col']` might be preferred over using `df.dropna()` <br><br>
    
3. Discuss why applying `del df['col']` before `df.dropna()` when both are used together could be important<br><br>
    
4. Remove all missing data from one of the datasets you're considering using some combination of `del df['col']` and/or `df.dropna()` and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.<br><br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Start a new ChatBot session **[but remember to first ask your ChatBot for summaries of your current session and perhaps coding results (so you can supply these in the homework as requested)]**, since your last ChatBot session has likely gotten quite long and has covered a lot of material at this point 
> - It can sometimes be helpful to reset ChatBot sessions to refocus them on the topics of inquiry without too much backlog history that might unintentionally bias things in certain directions and, of course, you can always re-introduce material from earlier conversations as it's relevant, such as for answering "D" based on reintroducing and updating code you made in a previous ChatBot session.  
> 
> #### ChatBot Scope Guidance
> 
> - This question is not interested in the general benefits of imputing missing data, or the general benefits of using `df.dropna()` and/or `del df['col']` to remove missing data, just how to most efficiently remove missing data if a user chooses to do so
> 
> - More sophisticated analyses for "filling in" rather than removing missing data (as considered here) are possible (based on making assumptions about missing data and using specific imputation methods or models) but these are "beyond the scope" of this homework assignment so this topics can be safely ignored for now
> 
> #### ChatBot Code Troubleshooting
> 
> A key issue to be aware of when asking ChatBots for help with something is that they are not running and checking code for correctess, and they often intertwine written instructions with code instructions; so, BEFORE YOU RUN ANY CODE provided by a ChatBot, you should check the following
> 
> 1. If this code changes an object or data, are you sure you want to run this code?
> 2. Can you easily "undo" the results of running code (e.g., from a copy `df_saved=df.copy()` or reloading the data) if running the code doesn't do what you want?
> 3. Is the state of the data what is expected by the code? Or have the objects been updated and changed so they're no longer what the code expects them to be? 
> 
> #### If you get any `Python` errors, copy and paste them into the ChatBot prompt and see if it can help you resove them; but, keep in mind the final point above becasue the ChatBot might not be aware of the state of your objects relative to the code it's producing...

</details><br>
    
#### 8. Give brief explanations in your own words for any requested answers to the questions below

> This problem will guide you through exploring how to use a ChatBot to troubleshoot code using the "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" data set 
> 
> To initialially constrain the scope of the reponses from your ChatBot, start a new ChatBot session with the following slight variation on the initial prompting approach from "2" above
> - "I am going to do some initial simple summary analyses on the titanic data set I've downloaded (https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv) which has some missing values, and I'd like to get your help understanding the code I'm using and the analysis it's performing"
        
1. Use your ChatBot session to understand what `df.groupby("col1")["col2"].describe()` does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you
    
> If needed, you can help guide the ChatBot by showing it the code you've used to download the data **AND provide it with the names of the columns** using either a summary of the data with `df.describe()` or just `df.columns` as demonstrated [here](../CHATLOG/COP/00017_copilot_groupby.md)
    
2. Assuming you've not yet removed missing values in the manner of question "7" above, `df.describe()` would have different values in the `count` value for different data columns depending on the missingness present in the original data.  Why do these capture something fundamentally different from the values in the `count` that result from doing something like `df.groupby("col1")["col2"].describe()`?

> Questions "4" and "6" above address how missing values are handled by `df.describe()` (which is reflected in the `count` output of this method); but, `count` in conjunction with `group_by` has another primary function that's more important than addressing missing values (although missing data could still play a role here).

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT<br><br>
    
    1. Forget to include `import pandas as pd` in your code 
       <br> 
       Use Kernel->Restart from the notebook menu to restart the jupyter notebook session unload imported libraries and start over so you can create this error
       <br><br>
       When python has an error, it sometimes provides a lot of "stack trace" output, but that's not usually very important for troubleshooting. For this problem for example, all you need to share with ChatGPT or search on google is `"NameError: name 'pd' is not defined"`<br><br>

    2. Mistype "titanic.csv" as "titanics.csv"
       <br> 
       If ChatBot troubleshooting is based on downloading the file, just replace the whole url with "titanics.csv" and try to troubleshoot the subsequent `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'` (assuming the file is indeed not present)
       <br><br>
       Explore introducing typos into a couple other parts of the url and note the slightly different errors this produces<br><br>
      
    3. Try to use a dataframe before it's been assigned into the variable
       <br> 
       You can simulate this by just misnaming the variable. For example, if you should write `df.groupby("col1")["col2"].describe()` based on how you loaded the data, then instead write `DF.groupby("col1")["col2"].describe()`
       <br><br>
       Make sure you've fixed your file name so that's not the error any more<br><br>
        
    4. Forget one of the parentheses somewhere the code
       <br>
       For example, if the code should be `pd.read_csv(url)` the change it to `pd.read_csv(url`<br><br>
        
    5. Mistype one of the names of the chained functions with the code 
       <br>
       For example, try something like `df.group_by("col1")["col2"].describe()` and `df.groupby("col1")["col2"].describle()`<br><br>
        
    6. Use a column name that's not in your data for the `groupby` and column selection 
       <br>
       For example, try capitalizing the columns for example replacing "sex" with "Sex" in `titanic_df.groupby("sex")["age"].describe()`, and then instead introducing the same error of "age"<br><br>
        
    7. Forget to put the column name as a string in quotes for the `groupby` and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question
       <br>
       For example, something like `titanic_df.groupby(sex)["age"].describe()`, and then `titanic_df.groupby("sex")[age].describe()`
        
#### 9. Have you reviewed the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?<br>
    
<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Just answering "Yes" or "No" or "Somewhat" or "Mostly" or whatever here is fine as this question isn't a part of the rubric; but, the midterm and final exams may ask questions that are based on the tutorial and lecture materials; and, your own skills will be limited by your familiarity with these materials (which will determine your ability to actually do actual things effectively with these skills... like the course project...)
    
</details>
    
***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***

# Recommended Additional Useful Activities [Optional]

The "Ethical Profesionalism Considerations" and "Current Course Project Capability Level" sections below **are not a part of the required homework assignment**; rather, they are regular weekly guides covering (a) relevant considerations regarding professional and ethical conduct, and (b) the analysis steps for the STA130 course project that are feasible at the current stage of the course<br><br>

<details class="details-example"><summary style="color:blue"><u>Ethical Professionalism Considerations</u></summary>

## Ethical Professionalism Considerations

> If the observed data is "no events occured" does this mean the data is "missing" and [should be ignored](https://priceonomics.com/the-space-shuttle-challenger-explosion-and-the-o)?
> 
> - NASA: \<determines temperature doesn't affects "o-ring" by subseting data to just "o-ring" incidents\>
> - Also NASA: \<launches the shuttle on a cold day\>

|No apparent "o-ring" failure and temperature relationship|Apparent between "o-ring" failure and temperature relationship|
|:-|:-|
if you just look at "o-ring" failure event data|if you instead look at ALL the data as you should|
|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image06-14.png)|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image02-33.png)|
|![](https://upload.wikimedia.org/wikipedia/commons/8/8b/Shuttle_Challenger_explosion.gif?20190203170223)|![](https://i.makeagif.com/media/10-04-2014/nT57xW.gif)|

<br>
    
</details>    

<details class="details-example"><summary style="color:blue"><u>Current Course Project Capability Level</u></summary>

## Current Course Project Capability Level

> The data we'll use for the STA130 course project is based on the [Canadian Social Connection Survey](https://casch.org/cscs). Please see the [data use agreement](https://static1.squarespace.com/static/60283c2e174c122f8ebe0f39/t/6239c284d610f76fed5a2e69/1647952517436/Data+Use+Agreement+for+the+Canadian+Social+Connection+Survey.pdf) regarding the appropriate and ethical professional use of this data (available at the bottom of the [CSCS](https://casch.org/cscs) webpage).
> 
> 1. Have a very quick look at the list of available variables available using the [link](https://drive.google.com/file/d/1ISVymGn-WR1lcRs4psIym2N3or5onNBi/view) (again at the bottom of the [CSCS](https://casch.org/cscs) webpage); then, 
> 2. examine the code in the first thirteen code cells of [STA130F24_CourseProject.ipynb](https://github.com/pointOfive/stat130chat130/blob/main/CP/STA130F24_CourseProject.ipynb) to get an initital understanding of how we might subset to different studies included in the [data](https://drive.google.com/file/d/1mbUQlMTrNYA7Ly5eImVRBn16Ehy9Lggo/view) (again accessible at the bottom of the [CSCS](https://casch.org/cscs) webpage); then,     
> 3. review the fourteenth and fifteenth cells (with the comments "Here's a high level summary of the data" and "And here are some explanations about the columns in the data") a little more closely to get a better sense of which columns seem to be the most interesting and whether or not they seem to have a lot of missing data
    
</details>        

### Afterward

Here are few ideas of some other kinds of interactions you might consider exploring with a ChatBot...

> While these are likely to be extremely practically valuable, they are not a part of the homework assignment, so do not include anything related to these in your homework submission

- With respect to improving ones ability in statistics, coding, communication, and other key data science skills
    - what is the ChatBots perception its own capabilities and uses as an AI-driven assistance tool 
    - and does ChatBots assessment of itself influence or agree with your own evalution of the ChatBot? 

- ChatBots can introduce and explain the "World War 2 planes" problem and the "Monte Hall" problem... 
    - how well does do they seem to do and introducing and explaining other "unintuitive surprising statistics paradoxes"?

- If you consider the process of writing about why you chose to take this course, and the skills you were hoping to build through this course with respect to your current ideas about what possible careers 
    - and how do you think the exercise would be different if you framed it as a dialogue with a ChatBot
    - and do you think the difference could be positive and productive, or potentially biasing and distracting?
    
- ChatBots sometimes immediately responds in simple helpful ways, but other times it gives a lot of extraneous information that can be overwheling... are you able to prompt and interact with ChatBots in manner that keeps its reponses helpful and focused on what you're interested in? 

- ChatBots tends to respond in a fairly empathetic and supportive tone...
    - do you find it helpful to discuss concerns you might have about succeeding in the course (or entering university more generally) with a ChatBot?
    
- For what purposes and in what contexts do you think a ChatBot could provide suggestions or feedback about your experiences that might be useful? 


make some changes


# 1. Using Animal Crossing data provided in the homework

# 2.1 print rows and columns in dataset

In [16]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)

# Get the number of rows and columns
num_rows, num_columns = villagers_df.shape
print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_columns}")
# chatbot link https://chatgpt.com/share/25448bde-7e05-4df7-b85f-ed204be3eccf

Number of rows: 391
Number of columns: 11


# 2.2 Observation & variable definitions

# Observation: 
An observation is a row of the dataset, which represents a single record. An observation typically contains values for multiple attributes (or features/variables), which are the different pieces of information being collected.
In the Animal Crossing dataset, an observation (row) represents a unique villager.

# Variable: 
A variable is a column in the dataset, representing a different characteristic/attribute of the villagers like the id, species, personality, etc.


# chatbot history + summary:
ChatGPT's definitions (chat history:https://chatgpt.com/share/8b3cf3fe-a46e-42c4-be71-f25bf6bae09c, https://chatgpt.com/share/00f930f1-af6b-4774-a7ad-02c80c184bc1) 
Here's a summary of our interaction:

1. You asked about what an **observation** is in the context of coding and datasets. I explained that an observation represents a single row in a dataset, with each row containing information about one data point (e.g., a villager in your Animal Crossing dataset).
   
2. You inquired about the definition of a **variable** in relation to your dataset. I clarified that a variable is a specific characteristic or attribute recorded for each observation (i.e., a column in your dataset), such as `name`, `species`, `personality`, etc.

We also discussed examples of categorical variables and how they are structured in your dataset.

# 3. summary of columns using df.describe() & df['col'].value_counts()

To get a summary of numeric columns, we should use villagers_df.describe(), which will provide us with the counts, mean, etc.
To get a summary of both numeric and non-numeric (categorical) columns, we should use villagers_df.describe(include='all'), which will summarize the count, unique values, most frequent value (top), and its frequency (freq).

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)
print(villagers_df.describe())
# summary of numerical columns

            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000


In [4]:
print(villagers_df.describe(include='all'))
# summary of ALL columns, including categorical ones

             row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391       

In [3]:
print(villagers_df.columns)

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')


In [5]:
print(villagers_df['row_n'].value_counts()) 

row_n
2      1
303    1
331    1
330    1
328    1
      ..
151    1
150    1
149    1
148    1
483    1
Name: count, Length: 391, dtype: int64


In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)
print(villagers_df['id'].value_counts()) 

id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64


# chatbot history + summary
chatbot history: https://chatgpt.com/share/8b3cf3fe-a46e-42c4-be71-f25bf6bae09c
couldn't get a summary for this chat because I went over the amount of questions you're allowed to ask. 

# 4. Explain discrepancies: compare df.shapes vs df.describe() using titanic dataset

# df.shape:
df.shape is an attribute (is a piece of data as opposed to performing an action and returning a result) containing the dimensions of the dataframe as (number of rows, number of columns). It is used to quickly understand the size of our dataframe.

df.shape just contains the dimensions of the dataset.

In the output of the cell below, the shape of the titanic dataset is (891, 15) which means that there are 891 rows and 15 columns of data in our table.

In [18]:
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
print("Shape of the dataset:", t_df.shape)

Shape of the dataset: (891, 15)


# df.describe():
df.describe() is a method (a function) which provides summary statistics of the dataframe's numerical columns by returning the count, mean, standard deviation, minimum, quartiles (25%, 50%, 75%), and maximum for each numerical column. It is used to quickly get a central tendency, dispersion, and shape of the numerical columns in a dataset.

By default, df.describe() returns numerical columns only. To return the counts and frequency statistics of non-numeric columns as well, we should use df.describe(include='all'). 

The statistics in df.describe() only reflect non-missing values. As a result, the count for some columns may be lower than the number of rows (from df.shape) if there are missing values. 
For example, in the 'age' column, 714 is displayed as the count, which represents how many people's ages are known of 891 rows total.

In [10]:
print(t_df.describe(include='all'))

          survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   8

# chatbot history + summary for question 4
https://chatgpt.com/share/f86b4698-8ac6-476c-97fe-9c05bd010e39
Here's a summary of our interaction:

1. **Question on `df.shape` and `df.describe()`**:
   - You asked about the difference between `df.shape` and `df.describe()`.
   - I explained that `df.shape` returns the dimensions of the DataFrame (as a tuple: number of rows, number of columns), while `df.describe()` provides summary statistics for the numerical columns of the DataFrame.

2. **Attributes vs Methods**:
   - You asked how these relate to attributes and methods.
   - I explained that:
     - **`df.shape` is an attribute** because it represents a property of the DataFrame and does not require parentheses.
     - **`df.describe()` is a method** because it performs a calculation and needs to be called with parentheses.

Let me know if you'd like more information or have any further questions!

# 5. difference between attributes and methods 

# Methods:
A method is a function that belongs to an object and performs an action,  computations, transformations, or other actions on the data. Methods require parentheses () to be performed.

df.describe() is an example of a method as it performs an operation on the DataFrame to calculate the summary statistics for the numerical columns, outputting the counts, mean, std, min, max, 25%, 50%, and 75% percentile. 
Since it’s a function, you need parentheses. 

# Attributes:
An attribute is a value associated with an object that can be referred to by a specific name. It is a piece of data that you should access it without using parentheses. 

df.shape is an example of an attribute. It represents the dimensions (number of rows, number of columns) of a dataframe. In terms of the Animal Crossing dataframe, df.shape outputs (891, 15), meaning the dataframe has 891 rows and 15 columns of data.


# function definition - programming vs mathematics:
# mathematics defintion:
A mathematical function is a relationship between a set of inputs and a set of possible outputs. Where each input gives exactly one output.

# programming definition:
A function (or method in python) in programming is a block of reusable code designed to perform a specific task. Taking inputs (called arguments or parameters) and returning a result or perform an action. In the case of df.describe(), the function (method) summarizes a the dataframes numerical data.

# chatbot history + summary:
https://chatgpt.com/share/f86b4698-8ac6-476c-97fe-9c05bd010e39
Here's a summary of our interaction:

1. **Question on `df.shape` and `df.describe()`**:
   - You asked about the difference between `df.shape` and `df.describe()`.
   - I explained that `df.shape` returns the dimensions of the DataFrame (as a tuple: number of rows, number of columns), while `df.describe()` provides summary statistics for the numerical columns of the DataFrame.

2. **Attributes vs Methods**:
   - You asked how these relate to attributes and methods.
   - I explained that:
     - **`df.shape` is an attribute** because it represents a property of the DataFrame and does not require parentheses.
     - **`df.describe()` is a method** because it performs a calculation and needs to be called with parentheses.

Let me know if you'd like more information or have any further questions!

https://chatgpt.com/share/aa9404ae-0bc8-4018-a766-97f5ff0d94e6 --> function in programming vs mathematics
Here’s a summary of our interaction:

You asked about the difference between a function in programming and a function in mathematics. I explained that:

- **Mathematical functions** define a relationship between inputs and outputs, where each input has a single output. They are deterministic, have no side effects, and are analyzed in terms of their abstract properties.
  
- **Programming functions** are reusable blocks of code that can perform tasks, return results, and sometimes cause side effects. They can be either deterministic or non-deterministic, handle complex logic, and interact with external states (e.g., files, variables).

I also highlighted the key differences in terms of purpose, behavior, and flexibility.
Let me know if you'd like more information or have any further questions!

# 6 definition of summary statistics

summary statistics are displayed when we call df.describe().
as mentioned above, by default, df.describe() returns numerical columns only. To return the counts and frequency statistics of non-numeric columns as well, we should use df.describe(include='all'). 

# using age as an example for each! 

# Count: 
The number of non-null observations in the dataset for each column. It shows how many valid non-null observations (rows) of data points are present.

For example, 150 indicates 150 valid age entries.

# Mean: 
The average value of the data in the column based solely on the available data that has valid numerical values. Any NaN (Not a Number) or null values are automatically excluded from the calculation of the mean.

It is calculated by summing all the non-null (non-missing) values in a dataset and then dividing by the number of those values. 

For example, the average age of villagers.

# Standard Deviation (std):
A measure of the amount of variation or dispersion in a set of values (dataset). It indicates how spread out the values are around the mean. 

A high standard deviation means the values are more spread out from the mean, while a low standard deviation means they are closer to the mean. 

For example, 10.2 indicates a deviation in age by 10.2 years from the mean.

# Minimum (min): 
The smallest value in the dataset in each column.
For example, 18 represents the youngest age recorded.

# Maximum (max): 
The largest value in the dataset in each column.
For example, 65 represents the oldest age recorded.

# 25th Percentile (25%): 
Also known as the first quartile, it represents the value below which 25% of the data falls. It’s the point at which 25% of the data is strictly less than this value and 75% is greater.
For example, 25 means that 25% of the ages are younger than 25 years.

# 50th Percentile (50%): 
Also known as the median, it’s the middle value of the dataset. Half of the data falls below this value, and half falls above it.
For example, 34 means half of the ages of villagers are 34 years or younger and half are older.

# 75th Percentile (75%): 

Also known as the third quartile, it represents the value below which 75% of the data falls. It’s the point at which 75% of the data is strictly less than this value and 25% is greater.
For example, 45 means that 75% of the ages are younger than 45 years.

# chatbot history + summary
https://chatgpt.com/share/39e3ea29-0001-4088-9408-de4cd9ef4610
Here’s a summary of our interaction:

1. **Definitions of Summary Statistics**: I provided explanations for the summary statistics (`count`, `mean`, `std`, `min`, `25%`, `50%`, `75%`, and `max`) generated by the `df.describe()` method in pandas.
   
2. **Explanation of Use Case**: I explained what a use case is, including its key components (actor, scenario, goal, preconditions, postconditions, and exceptions) and gave an example of a use case for "User Login."

3. **Mean Calculation**: I explained that the mean is calculated by summing all non-null values in a dataset and dividing by the number of those values, and that `NaN` values are excluded from this calculation.

If there's anything else you'd like to know or clarify, just let me know!
Let me know if you'd like to dive deeper into any of these topics!

# 7.  when to use df.dropna() vs del df['col']

# df.dropna()
a function that removes a single value in a row with any missing values. 
df.dropna() is used when dealing with rows or columns that contain missing values and you want to clean your dataset by removing incomplete data based on how many missing values you have. This helps ensure that your analysis is based on complete (non-missing) data points. 

# del df['col'] 
deletes an entire column based on an extreme amount of missing values. 
del df['col'] is used when you want to remove specific columns that you have deemed unnecessary for your analysis. This is not about cleaning missing values but about simplifying the structure of your dataset.

In the context of the Animal Crossing dataset, the decision to use df.dropna() versus del df['col'] depends on what you're trying to accomplish with the data, especially in terms of handling missing values and column deletion.

# 7.1 use case for when df.dropna() is preferred over del df['col']
df.dropna() is used when we're interested in cleaning data by eliminating rows or columns based on the presence of missing values.

df.dropna() is commonly used to clean data, prepare it for machine learning algorithms (which require a complete dataset, meaning no missing/null values), to simplify analysis, and to ensure consistency with data collection. 

For example, if we want to clean our data by dropping rows where any villager's details (such as species, personality, or birthday) have missing values (NaNs), we'd use df.dropna().

In [1]:
# example of using df.dropna() with Animal Crossing dataset
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)
# Drop rows with missing values in any column
villagers_df.dropna()

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[379 rows x 11 columns]

When a column is essential for your analysis (like species or personality) and missing data would skew your results, use dropna(subset=[columns]).
See cell below for an example of cleaning rows

In [7]:
# cleaning rows with missing values for specific analysis which requires a complete data (no missing values)
# use df.dropna() to drop rows that contains missing values as it might cause issues
villagers_df.dropna(subset=['species', 'personality', 'birthday'])

row_n      name  gender    species birthday personality             song  \
0        2   Admiral    male       bird     1-27      cranky       Steep Hill   
1        3   Agent S  female   squirrel      7-2       peppy          DJ K.K.   
2        4     Agnes  female        pig     4-21        uchi       K.K. House   
3        6        Al    male    gorilla    10-18        lazy       Steep Hill   
4        7   Alfonso    male  alligator      6-9        lazy      Forest Life   
..     ...       ...     ...        ...      ...         ...              ...   
386    475    Winnie  female      horse     1-31       peppy         My Place   
387    477  Wolfgang    male       wolf    11-25      cranky        K.K. Song   
388    480      Yuka  female      koala     7-20      snooty     Soulful K.K.   
389    481      Zell    male       deer      6-7        smug         K.K. D&B   
390    483    Zucker    male    octopus      3-8        lazy  Spring Blossoms   

       phrase            full_id  \
0     aye aye   villager-admiral   
1    sidekick   villager-agent-s   
2     snuffle     villager-agnes   
3     Ayyeeee        villager-al   
4    it'sa me   villager-alfonso   
..        ...                ...   
386    hay-OK    villager-winnie   
387   snarrrl  villager-wolfgang   
388   tsk tsk      villager-yuka   
389     pronk      villager-zell   
390     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 10 columns]

For analyzing complete subsets of the data, such as song-based analysis, use dropna() on relevant columns to ensure no rows with critical missing data distort your results by emoving the columns with excessive missing data that aren't contributing to your analysis.

see cell below for an example of removing rows   

In [9]:
# removing rows with missing data (i.e. 'song')
# use df.dropna() to drop missing values in 'song' column that could make results incomplete 
villagers_df.dropna(subset=['song'])

row_n      name  gender    species birthday personality             song  \
0        2   Admiral    male       bird     1-27      cranky       Steep Hill   
1        3   Agent S  female   squirrel      7-2       peppy          DJ K.K.   
2        4     Agnes  female        pig     4-21        uchi       K.K. House   
3        6        Al    male    gorilla    10-18        lazy       Steep Hill   
4        7   Alfonso    male  alligator      6-9        lazy      Forest Life   
..     ...       ...     ...        ...      ...         ...              ...   
386    475    Winnie  female      horse     1-31       peppy         My Place   
387    477  Wolfgang    male       wolf    11-25      cranky        K.K. Song   
388    480      Yuka  female      koala     7-20      snooty     Soulful K.K.   
389    481      Zell    male       deer      6-7        smug         K.K. D&B   
390    483    Zucker    male    octopus      3-8        lazy  Spring Blossoms   

       phrase            full_id  \
0     aye aye   villager-admiral   
1    sidekick   villager-agent-s   
2     snuffle     villager-agnes   
3     Ayyeeee        villager-al   
4    it'sa me   villager-alfonso   
..        ...                ...   
386    hay-OK    villager-winnie   
387   snarrrl  villager-wolfgang   
388   tsk tsk      villager-yuka   
389     pronk      villager-zell   
390     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[380 rows x 10 columns]

# 7.2 use case for when del df['col'] is preferred over df.dropna()
del df['col'] is used when you know exactly which column you want to remove from the DataFrame, regardless of whether it contains missing values. This is helpful when aspecific column is irrelevant to your analysis, and you want to remove it for clarity and focus. 

For example, if the 'id' column isn't relevant in my analysis of 'birthday' and 'songs', then I can delete that column using del df['col'] to simplify the data I use for my analysis. 
An example of deleting the 'id' column is shown in the cells below.

In [2]:
# before deleting a column (using 'id' as the example)
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)
villagers_df.columns

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')

In [3]:
# deleting the 'id' column
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)
del villagers_df['id']

In [5]:
# Display columns after deletion
print("Columns after deletion:", villagers_df.columns)

Columns after deletion: Index(['row_n', 'name', 'gender', 'species', 'birthday', 'personality', 'song',
       'phrase', 'full_id', 'url'],
      dtype='object')


# 7.3 applying del df['col'] before df.dropna() is important

When using df.dropna() and deleting a column with del df['col'], the order matters. If you delete the column first del df['col'], you ensure that you don't unintentionally remove important rows due to NaN values in that column. Then, when you apply df.dropna(), it only affects the remaining columns. 

If you use df.dropna() before deleting the column, it will remove rows with NaN values in any column, including the one you might want to delete later. Thus, you might end up dropping rows you could have kept if you deleted the column first.

Deleting unnecessary columns before applying df.dropna() can help maintain data integrity and improve performance by reducing the amount of data to process in subsequent steps.

# chatbot history + summary
https://chatgpt.com/share/6f6f246f-3654-4628-99aa-0747122d0dc7
You asked about the importance of applying `del df['col']` before `df.dropna()` when both are used together. I explained that deleting the column first ensures you don’t inadvertently drop rows that might have been important if they had NaN values only in the column you plan to remove. This approach helps maintain data integrity and can improve performance by reducing computational overhead.

# 7.4 how to efficiently remove missing data using del df['col'] df.dropna() 

In [11]:
# see how many missing values in the animal crossing dataset
import pandas as pd
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
villagers_df = pd.read_csv(url)
# Assuming df is your DataFrame
missing_values_villagers = villagers_df.isna().sum()
print(missing_values_villagers)

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


In [13]:
villagers_df['id'].count()

390

In [14]:
villagers_df['song'].count()

380

there is 1 missing value from 'id' and 11 missing values from song. 
therefore, the efficient way to approach removing the null values is to use df.dropna() to drop the 1 null value from the 'id' column in which we would only lose 0.2% of data in the entire dataset.

with only 2.89% missing values, deleting the entire 'song' column may be too drastic unless the column is irrelevant to your analysis. 
if it is irrelavent to our analysis, we can use del df['song'] to delete the 'song' column.
if it is relevant to the analysis, then we would be loosing 97.11% of non-missing vlaues if we deleted the entire column. 
in the case that we want to keep as much non-missing values in our dataset (efficiently removing missing data) we can use df.dropna() again for the song column.

In [10]:
# remove rows with missing data
import pandas as pd
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
villagers_df = pd.read_csv(url)

# delete

# Drop rows with any missing data
villagers_df.dropna()

# Drop columns with any missing data
villagers_df.dropna(axis=1)

row_n      name  gender    species birthday personality    phrase  \
0        2   Admiral    male       bird     1-27      cranky   aye aye   
1        3   Agent S  female   squirrel      7-2       peppy  sidekick   
2        4     Agnes  female        pig     4-21        uchi   snuffle   
3        6        Al    male    gorilla    10-18        lazy   Ayyeeee   
4        7   Alfonso    male  alligator      6-9        lazy  it'sa me   
..     ...       ...     ...        ...      ...         ...       ...   
386    475    Winnie  female      horse     1-31       peppy    hay-OK   
387    477  Wolfgang    male       wolf    11-25      cranky   snarrrl   
388    480      Yuka  female      koala     7-20      snooty   tsk tsk   
389    481      Zell    male       deer      6-7        smug     pronk   
390    483    Zucker    male    octopus      3-8        lazy     bloop   

               full_id                                                url  
0     villager-admiral  https://villagerdb.com/images/villagers/thumb/...  
1     villager-agent-s  https://villagerdb.com/images/villagers/thumb/...  
2       villager-agnes  https://villagerdb.com/images/villagers/thumb/...  
3          villager-al  https://villagerdb.com/images/villagers/thumb/...  
4     villager-alfonso  https://villagerdb.com/images/villagers/thumb/...  
..                 ...                                                ...  
386    villager-winnie  https://villagerdb.com/images/villagers/thumb/...  
387  villager-wolfgang  https://villagerdb.com/images/villagers/thumb/...  
388      villager-yuka  https://villagerdb.com/images/villagers/thumb/...  
389      villager-zell  https://villagerdb.com/images/villagers/thumb/...  
390    villager-zucker  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 9 columns]

# chatbot history + summary
https://chatgpt.com/share/4a81a680-2697-4c22-9fc1-d504e9e02706
We discussed handling missing values in the `song` column of your villagers dataset, considering whether to drop rows, impute, or delete the column.

https://chatgpt.com/share/1213ae67-7f37-42a1-983a-2adba5727130 

Here's a summary of our interaction so far:

    Introduction to df.dropna():
        We discussed when to use df.dropna() for cleaning rows or columns that contain missing values. This is useful when preparing datasets for analysis or ensuring completeness in machine learning models.

    Context of Your Dataset:
        Your dataset from TidyTuesday contains information on villagers. We explored how df.dropna() can be used in this context to remove rows or columns with missing values in specific fields like species, personality, or birthday.

    Examples of Using df.dropna() in Your Dataset:
        Drop rows with missing values in key columns (e.g., species, personality, birthday).
        Remove columns with excessive missing data, such as dropping a column with too many NaNs (e.g., catchphrase).
        Drop columns based on a threshold (e.g., if more than 50% of a column's values are missing).

    When to Use del df['col']:
        We explored when to use del df['col'] to remove irrelevant or redundant columns:
            Remove irrelevant columns (e.g., file_name if you don’t need file paths for analysis).
            Remove redundant columns (e.g., birthday_month if the full birthday column is already present).
            Remove columns with privacy concerns (e.g., real_name).
            Remove columns with excessive missing values or that don’t contribute to your analysis (e.g., catchphrase if it’s missing a lot of data).
            Simplify the dataset to focus on key attributes (e.g., removing gender if it’s irrelevant to your analysis).

Would you like further clarification on any of these topics or specific assistance with your dataset?


# 8.1 demonstrate what groupby does
The df.groupby("col1")["col2"].describe() function groups the data by the unique values in the column col1 and selects the col2 values. It then generates summary statistics for the selected column, col2, such as count, mean, standard deviation, min, max, 25%, 50%, and 75% percentiles. 

In the example in the cell below summary statistics for 'age' are calculated seperately for each group of col 1 'survived'.

In [31]:
import pandas as pd
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
t_df = pd.read_csv(url)

#Group by 'survived' and describe the 'age' column. 0 means did not survive, 1 means survived
t_df.groupby('survived')['age'].describe()

count       mean        std   min   25%   50%   75%   max
survived                                                           
0         424.0  30.626179  14.172110  1.00  21.0  28.0  39.0  74.0
1         290.0  28.343690  14.950952  0.42  19.0  28.0  36.0  80.0

424 passengers did not survive, the mean ages of the passengers who didn't survive is 30.6yrs, with a std of 14.2yrs, minimum age of 1, max age of 74, 25% of the passengers who didn't survive were younger than 21, 75% of the passengers who didn't survive were over 39.

290 passengers did survive, with a mean age of 28.3yrs, std of 14.95yrs, minimum age of 0.42yrs, max age of 80yrs, 25% of survivors were younger than 19, 75% of survivors were older than 36.

# chatbot history + summary
https://chatgpt.com/share/a805d037-609f-42b9-823e-d21f5286b82e
We discussed the differences between `df.describe()` and `df.groupby().describe()` in pandas. `df.describe()` provides summary statistics for the entire DataFrame, while `df.groupby().describe()` gives summary statistics for each group defined by a `groupby` operation. We then focused on the output of `t_df.groupby('survived')['age'].describe()`, explaining that percentiles (25th, 50th, and 75th) indicate the distribution of ages within each survival group, helping to understand if age influenced survival rates.

# 8.2 df.describe() vs df.groupby('col1')['col2'].describe()

the count values produced by df.groupby("col1")["col2"].describe() and df.describe() are different because of how the data is treated in each case. both functions ignore the null/NaN values and provide summary statistics.

df.groupby('col1')['col2'].describe function is more specific as it groups the count by comparing the non-missing values of one column with the unique values in another column.

df.describe counts the non-missing values of the entire dataset rather than specific columns.

# value counts using df.describe() 
df.describe() performs a calculation is done based on the entire dataset. It counts the number of non-missing values across the entire column and does not consider any grouping. This function creates one set of summary statistics for the whole dataset.

In [8]:
t_df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [11]:
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.describe() 
# The output summary statistics for the entire dataset, without grouping

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

# value counts using df.groupby("col1")["col2"].describe()
the count values produced by df.groupby("col1")["col2"].describe() are made by grouping the data based on the number of non-missing (NaN) values in col2 for each unique group in col1 to calculate the summary statistics within each group. 

unlike df.describe() which provides overall summary statistics for the entire dataset, df.groupby('col1')['col2'].describe() gives group-specific statistics (e.g., mean, median, min, max, etc.) for col2 within each group of col1. 

by focusing on counting non-missing values within a each group, it provides a count per group. 

this helps compare the statistics across different groups.

in the cell below, you can see how df.groupby("col1")["col2"].describe() shows how many valid survivors values are there for each gender in the Titanic dataset.

In [20]:
t_df.groupby('sex')['survived'].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

Count for females = 314 shows that there are 314 non-missing age values for female passengers.
Count for males = 577 shows that there are 577 non-missing age values for male passengers.

# how the counts differ 
in df['age'].describe(), the count is 714 because it considers the total number of non-missing age values across all passengers.

in df.groupby('sex')['survived'].describe(), the grouped count differs from df.describe() because the counts are now being split between the two groups 'sex' and 'survived' where each group's count reflects the number of non-missing age values within that group (165 females and 360 males).

# chatbot history + summary
https://chatgpt.com/share/9acfd934-c5de-4fc8-873c-9a4704bae3f2
We discussed how `df.groupby('col1')['col2'].describe()` works in the context of the Titanic dataset and explored how it differs from `df.describe()`. We covered:

1. **`groupby()` vs `describe()`**:
   - `groupby().describe()` provides detailed summary statistics (mean, median, quartiles, etc.) for each group (e.g., `sex`), while `describe()` gives overall statistics for the entire dataset or a column.
   - `groupby().count()` shows non-missing values within each group, while `describe()` provides statistics for the whole dataset, ignoring missing values.
   
2. **Advantages of `groupby().describe()`**:
   - It provides **group-specific statistics**, reveals **patterns across groups**, and handles missing values automatically, making it great for exploratory data analysis.

The key takeaway is that `groupby().describe()` offers more detailed insights into your dataset by allowing you to compare statistics across different groups.

# 8.3 make errors on purpose, use chat and google to troubleshoot; determine which is better at troubleshooting

# A

In [3]:
# forget to include import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.describe()

NameError: name 'pd' is not defined

In [4]:
# fix the issue above by importing pandas as pd
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

- chat took time to generate, google search loaded faster

# B

In [4]:
# type file name wrong
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
t_df = pd.read_csv(t_url)
t_df.shape()

HTTPError: HTTP Error 404: Not Found

In [5]:
# fix issue above by checking for typographical errors (mispelling file names)
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

- chat offered multiple different solutions to different sources of problems that could cause the error message for B
- google search directed me to stackoverflow which offered me one solution to the exact problem in B: mispelling the file name

# C

In [6]:
# use a dataframe that hasn't been assigned/defined yet

DF.groupby("col1")["col2"].describe()

NameError: name 'DF' is not defined

In [7]:
# fix by using a defined dataframe like 't_df' or alternatively defining the variable 'DF' 
# i will be using 't_df' since i've defined it previously
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

- chat offered multiple different solutions to different sources of problems that could cause the error message for c
- google search directed me to stackoverflow which offered me one solution to the exact problem in c: using an undefined df

# D

In [8]:
# forget a parentheses
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url
t_df.describe()

SyntaxError: '(' was never closed (3285048580.py, line 4)

In [9]:
# fix by checking for missing characters (i.e. parentheses)
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

- chat offered multiple different solutions to different sources of problems that could cause the error message for D
- google search directed me to stackoverflow which offered me one solution to the exact problem in D: forgetting a parentheses

# E

In [13]:
# mistype one of the names of the chained functions with the code
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.groupby("sex")["age"].describleyaya()

AttributeError: 'SeriesGroupBy' object has no attribute 'describleyaya'

In [14]:
# fix by checking for typographical errors
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

- chat took time to generate, gave me mutliple different solutions and examples
- google search loaded faster, provided one solution that was specific to my error

# F

In [15]:
# use a column name that's not in the titanic dataset, python is case sensitive so adding a capital-letter will cause a problem
# 'Sex' instead of 'sex'
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

In [16]:
# fix by changing 'Sex' which isn't a column name in the dataset to 'sex' which is a column name in the dataset
# since python is case sensitive, this caused an error when running the code
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [17]:
# same thing but using 'Age' instead of 'age'
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.groupby("sex")["Age"].describe()

KeyError: 'Column not found: Age'

In [18]:
# fix by changing 'Age' which isn't a column name in the dataset to 'age' which is a column name in the dataset
# since python is case sensitive, this caused an error when running the code
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

- chat's loading time was similar to google search
- both gave me the same solution to the error in F

# G

In [19]:
# forget to put the column name as a string
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.groupby("sex")[age].describe()

NameError: name 'age' is not defined

In [20]:
# fix by making the column name into a string so the function can be called
import pandas as pd
t_url= "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
t_df = pd.read_csv(t_url)
t_df.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

- both chatGPT & google search gave me similiar wokring code in comparable loading times
- however, stackoverflow (a search result) gave me a better reason for why forgetting to put the column name as a string will produce this error 

# conclusion: chat or google (stackoverflow)
In my opinion, both ChatGPT and Google search were equally fast in giving me a solution, although I did spend a decent amount of time cross-checking every answer from ChatGPT with code/solutions people have posted on sites like StackOverflow.

On sites like StackOverflow or Medium, there are comments and you can see users' opinions or tips/tricks for troubleshooting and understanding the problems I had. This makes me trust these solutions from Google search more than blindly executing code from ChatGPT, which needed trial and error to see if the generated code will actually help.


# chatbot history + summaries:
https://chatgpt.com/share/bb07f3bc-ebc2-42fd-b6ae-365fd283ea67
Here's a brief summary of our interaction:

1. **Error Handling**:
   - **`NameError: name 'pd' is not defined`**: You needed to import the `pandas` library using `import pandas as pd`.
   - **`NameError: name 'titanics' is not defined`**: There was a typo; it should be `t_df`, not `titanics`.
   - **`HTTPError: HTTP Error 404: Not Found`**: The URL for the Titanic dataset was incorrect or the file was not available.
   - **`SyntaxError: '(' was never closed`**: There was an unmatched parenthesis in the `pd.read_csv()` function call.
   - **`KeyError` issues**: Incorrect column names were used; column names need to be verified and matched correctly.
   - **`AttributeError`**: A typo in method name (`describleyaya` instead of `describe()`).
   - **`NameError: name 'age' is not defined`**: Column names should be in quotes when referencing them.

2. **Solutions Provided**:
   - Ensured correct import statements and function usage.
   - Corrected column names and method names based on dataset structure.
   - Recommended checking column names using `print(t_df.columns)` to resolve issues.

Let me know if there's anything else you need help with!

# 9 
yes i have taken a look at the wiki-textbook and piazza.

# other chatbot history links and summaries
https://chatgpt.com/share/1213ae67-7f37-42a1-983a-2adba5727130 --> df.dropna vs del df['col']

Here's a summary of our interaction:

1. **Question on `df.shape` and `df.describe()`**:
   - You asked about the difference between `df.shape` and `df.describe()`.
   - I explained that `df.shape` returns the dimensions of the DataFrame (as a tuple: number of rows, number of columns), while `df.describe()` provides summary statistics for the numerical columns of the DataFrame.

2. **Attributes vs Methods**:
   - You asked how these relate to attributes and methods.
   - I explained that:
     - **`df.shape` is an attribute** because it represents a property of the DataFrame and does not require parentheses.
     - **`df.describe()` is a method** because it performs a calculation and needs to be called with parentheses.

Let me know if you'd like more information or have any further questions!

Here's a summary of our interaction so far:

1. You mentioned you are working with a dataset about characters from *Animal Crossing* from the TidyTuesday project (specifically from May 5, 2020).
2. You asked for details on the dataset structure, including how many rows and columns it has.
3. I initially tried to load and analyze the dataset but was unable to access the tool for more advanced operations. I then provided you with a Python code snippet to find the number of rows and columns in the dataset yourself.

Let me know if you'd like further assistance!

Here's a summary of our interaction:

1. **Dataset Exploration**:
   - You are working with a dataset from the TidyTuesday project (May 5, 2020), focused on villagers.
   - We discussed key terms like **observations** (rows) and **variables** (columns), focusing on how they relate to your dataset.

2. **Inspecting the Dataset**:
   - You asked about the size of the dataset, which we explored using `df.shape` (which gave 391 rows and 11 columns).
   - We also examined how `df.describe()` provides a summary of the numeric columns by default and how it differs from `df.shape` in terms of analyzed columns and reported counts (non-missing values).

3. **Basic Data Summary**:
   - I provided steps to check for missing values, display the column names, and perform simple summarizations like `df.describe(include='all')`, `df.info()`, and `df.isnull().sum()`.

4. **Comparing `df.shape` vs. `df.describe()`**:
   - We discussed how `df.shape` reports the total size, while `df.describe()` reports only non-missing values and typically only analyzes numeric columns unless specified otherwise (`include='all'`).

Throughout our interaction, we focused on understanding the structure and content of the villagers dataset, preparing you to summarize and analyze it efficiently.

Let me know if you'd like further assistance!